src/profile-logic/import/android-trace.js

```diff
@@ -0,0 +1,668 @@
+      if (currentStack === null) {
+        // This has been observed to happen in tracing-based traces (rather than sampling-based traces). Not sure why.
+        // console.warn('exiting method when stack is empty');
+        return;
+      }
+      currentStack = stackTable.data[currentStack][1];
+    }
+
+    // Called before enter/exitMethod are called for this time
+    function advanceTimeTo(timestampInMSSinceStartTime) {
+      if (nextSampleTimestamp === null) {
+        nextSampleTimestamp = timestampInMSSinceStartTime;
+        return;
+      }
+
+      while (nextSampleTimestamp < timestampInMSSinceStartTime) {
+        samples.data.push([currentStack, nextSampleTimestamp]);
+        if (honorOriginalSamplingTimestamps) {
+          // Only use this loop to fill up any gaps that are at least 2 * interval wide.
+          // When less than 2 * interval remains, snap to the original next timestamp.
+          if (
+            timestampInMSSinceStartTime - nextSampleTimestamp <
+            2 * intervalInMsec
+          ) {
+            nextSampleTimestamp = timestampInMSSinceStartTime;
+            break;
+          }
+        }
+        nextSampleTimestamp += intervalInMsec;
+      }
+    }
+
+    function finish() {
+      if (nextSampleTimestamp !== null) {
+        samples.data.push([currentStack, nextSampleTimestamp]);
+      }
+      return {
+        tid,
+        pid,
+        name,
+        registerTime: name === 'main' ? 0 : samples.data[0][1],
+        unregisterTime: null,
+        markers,
+        samples,
+        frameTable,
+        stackTable,
+        stringTable,
+      };
+    }
+
+    return {
+      enterMethod,
+      exitMethod,
+      advanceTimeTo,
+      finish,
+    };
+  }
+
+  const threadMap = new Map();
+
+  if (methodActions.length > 0) {
+    for (let i = 0; i < methodActions.length; i++) {
+      const { tid, methodId, globalTime, action } = methodActions[i];
+      let thread = threadMap.get(tid);
+      if (thread === undefined) {
+        const traceThread = threads.find(t => t.tid === tid);
+        if (!traceThread) {
+          throw new Error(
+            `Encountered method action for unknown thread ${tid}`
+          );
+        }
+        thread = _createThread(
+          traceThread.threadName,
+          headerDetails.pid ? +headerDetails.pid : 1,
+          tid
+        );
+        threadMap.set(tid, thread);
+      }
+      thread.advanceTimeTo(globalTime / 1000);
+      switch (action) {
+        case 'enter':
+          thread.enterMethod(methodId);
+          break;
+        case 'exit':
+        case 'exit-unroll':
+          thread.exitMethod(methodId);
+          break;
+        default:
+          throw new Error(`Unexpected action ${action}`);
+      }
+    }
+  }
+
+  const threadArray = Array.from(threadMap.values()).map(thread =>
+    thread.finish()
+  );
+
+  return {
+    meta: {
+      interval: intervalInMsec,
+      processType: 0,
+      product: 'Firefox',
```

Comment:
Maybe instead of the product being "Firefox", this could be the imported format, e.g. "Android CPU Trace".

Reply:
I've changed it to "ART Trace (Android)". The name "ART trace" comes from [this comment in the Android Studio code](https://android.googlesource.com/platform/tools/adt/idea/+/refs/heads/studio-master-dev/profilers/src/com/android/tools/profilers/cpu/CpuCaptureParser.java#241).